src/main.cpp

```diff
@@ -3855,13 +3855,19 @@ bool CheckBlockHeader(
     const CChainParams& chainparams,
     bool fCheckPOW)
 {
+    auto consensusParams = chainparams.GetConsensus();
+
     // Check block version
     if (block.nVersion < MIN_BLOCK_VERSION)
         return state.DoS(100, error("CheckBlockHeader(): block version too low"),
                          REJECT_INVALID, "version-too-low");
 
-    // Check Equihash solution is valid
-    if (fCheckPOW && !CheckEquihashSolution(&block, chainparams.GetConsensus()))
+    // Check Equihash solution is valid. The main check is in ContextualCheckBlockHeader,
+    // because we currently need to know the block height. That function skips the genesis
+    // block because it has no previous block, so we check it specifically here.
+    if (fCheckPOW &&
+        block.GetHash() == consensusParams.hashGenesisBlock &&
+        !CheckEquihashSolution(&block, 0, consensusParams))
```

Comment:
I'm concerned this may have unintended effects. Specifically, checking the proof of work in block headers (which we may have even when we don't have the full block) might be required to prevent some DoS attacks.

Reply:
We do still check the full proof-of-work in block headers, regardless of having the full block. `AcceptBlockHeader` calls `CheckBlockHeader` followed by `ContextualCheckBlockHeader`. See also https://github.com/zcash/zcash/pull/4451#issuecomment-613966669.

Reply:
OK cool, will re-review. Sorry for the false alarm (but please let's follow the review policy, since this change very clearly was nontrivial enough to need it!)